crates/wasi/src/preview2/preview2/filesystem.rs

```diff
@@ -837,10 +828,112 @@ impl<T: WasiView> filesystem::Host for T {
+        let meta_b = get_descriptor_metadata(table, b).await?;
+        if meta_a.dev() == meta_b.dev() && meta_a.ino() == meta_b.ino() {
+            // MetadataHashValue does not derive eq, so use a pair of
+            // comparisons to check equality:
+            debug_assert_eq!(
+                calculate_metadata_hash(&meta_a).upper,
+                calculate_metadata_hash(&meta_b).upper
+            );
+            debug_assert_eq!(
+                calculate_metadata_hash(&meta_a).lower,
+                calculate_metadata_hash(&meta_b).lower
+            );
+            Ok(true)
+        } else {
+            // Hash collisions are possible, so don't assert the negative here
+            Ok(false)
+        }
+    }
+    async fn metadata_hash(
+        &mut self,
+        fd: types::Descriptor,
+    ) -> Result<types::MetadataHashValue, types::Error> {
+        let table = self.table();
+        let meta = get_descriptor_metadata(table, fd).await?;
+        Ok(calculate_metadata_hash(&meta))
+    }
+    async fn metadata_hash_at(
+        &mut self,
+        fd: types::Descriptor,
+        path_flags: types::PathFlags,
+        path: String,
+    ) -> Result<types::MetadataHashValue, types::Error> {
+        let table = self.table();
+        let d = table.get_dir(fd)?;
+        // No permissions check on metadata: if dir opened, allowed to stat it
+        let meta = d
+            .spawn_blocking(move |d| {
+                if symlink_follow(path_flags) {
+                    d.metadata(path)
+                } else {
+                    d.symlink_metadata(path)
+                }
+            })
+            .await?;
+        Ok(calculate_metadata_hash(&meta))
+    }
+}
+
+async fn get_descriptor_metadata(
+    table: &Table,
+    fd: types::Descriptor,
+) -> Result<cap_std::fs::Metadata, types::Error> {
+    if table.is_file(fd) {
+        let f = table.get_file(fd)?;
+        // No permissions check on metadata: if opened, allowed to stat it
+        Ok(f.spawn_blocking(|f| f.metadata()).await?)
+    } else if table.is_dir(fd) {
+        let d = table.get_dir(fd)?;
+        // No permissions check on metadata: if opened, allowed to stat it
+        Ok(d.spawn_blocking(|d| d.dir_metadata()).await?)
+    } else {
+        Err(ErrorCode::BadDescriptor.into())
+    }
+}
+
+fn calculate_metadata_hash(meta: &cap_std::fs::Metadata) -> types::MetadataHashValue {
+    use cap_fs_ext::MetadataExt;
+    // Without incurring any deps, std provides us with a 64 bit hash
+    // function:
+    use std::hash::{BuildHasher, Hasher};
+    // We only want to initialize the BuildHasher impl once so that each Hasher we create performs
+    // the same calculation.
+    // Note that this means that the metadata hash (which becomes a preview1 ino) will be different
+    // in each wasmtime process!
+    static BUILD_HASHER: std::sync::OnceLock<std::collections::hash_map::RandomState> =
+        std::sync::OnceLock::new();
```

Comment:
Why not use `std::collections::hash_map::DefaultHasher` here instead?

Reply:
I missed it, but that is a better option.